### ISSUE TITLE
Infer static return type from static::class

### ISSUE THREAD

**Issue by BenMorel** [vimeo/psalm#10309]
Below is a simplified example from [brick/math](https://github.com/brick/math).

I have the following hierarchy of classes:

- `BigNumber`
  - `BigInteger`
  - `BigDecimal`

To get an instance, you can call the `of()` factory method on any of the classes:

- `BigNumber::of()` returns a `BigInteger` or a `BigDecimal` depending on the input value
- `BigInteger::of()` and `BigDecimal::of()` ensure that you get an instance of themselves

I'm therefore attempting to use the `static` type as return value of `of()`, but am getting this Psalm error:

> ERROR: LessSpecificReturnStatement - src/BigNumber.php:61:20 - The type 'Brick\Math\BigNumber' is more general than the declared return type 'Brick\Math\BigNumber&static' for Brick\Math\BigNumber::of (see https://psalm.dev/129)

The issue is on this code block:

```php
if (static::class === BigNumber::class) {
    return $value;
}
```

where Psalm does not seem to understand that the return value (of type `BigNumber`) passes the `static` return type check when `of()` is called on `BigNumber`.

Reproduce: https://psalm.dev/r/df90df8579

---

```php
abstract class BigNumber
{
    public static function of(BigNumber|int|float|string $value) : static
    {
        $value = self::_of($value);

        if (static::class === BigNumber::class) {
            return $value;
        }

        return static::from($value);
    }

    private static function _of(BigNumber|int|float|string $value) : BigNumber
    {
        if ($value instanceof BigNumber) {
            return $value;
        }

        if (is_int($value)) {
            return new BigInteger();
        }

        return new BigDecimal();
    }

    abstract protected static function from(BigNumber $number): static;
}

final class BigInteger extends BigNumber
{
    protected static function from(BigNumber $number): static
    {
        // actually attempts a conversion to BigInteger
        return new BigInteger();
    }
}

final class BigDecimal extends BigNumber
{
    protected static function from(BigNumber $number): static
    {
        // actually attempts a conversion to BigDecimal
        return new BigDecimal();
    }
}

```

**Comment by danog** [comment]
A simpler reproduction case: https://psalm.dev/r/cf41f76b20

Basically a new `self::class === static::class` assertion should be introduced internally in Psalm to fix.

As a workaround, you can use an explicit assertion: https://psalm.dev/r/f3bb64fc23

**Comment by BenMorel** [comment]
@danog Thanks for the simpler reproduction case. I'm a bit confused by your `AssertionError("impossible")` though, this line is actually triggered if you execute your example: https://3v4l.org/4lZt4

It is triggered when `test()` is called on a subclass of `a`.

> As a workaround, you can use an explicit assertion

Good point!

**Comment by danog** [comment]
>this line is actually triggered if you execute your example

Yep, that's kind of the point, this way the return type of the function is respected, one way or another :)

**Comment by BenMorel** [comment]
Oh, sorry, I misunderstood the meaning then!
